ydb/library/yql/minikql/comp_nodes/mkql_wide_top_sort.cpp

```diff
@@ -208,20 +352,303 @@ using TBase = TComputationValue<TState<HasCount>>;
+
+    enum class EOperatingMode {
+        InMemory,
+        Spilling,
+        ProcessSpilled
+    };
+
+    size_t GetStorageSize() const {
+        return std::max<size_t>(Count << 2ULL, 1ULL << 8ULL);
+    }
+
+    void ResetFields() {
+        NUdf::TUnboxedValuePod* ptr;
+        if constexpr (!HasCount) {
+            auto pos = Storage.size();
+            Storage.insert(Storage.end(), Indexes.size(), {});
+            ptr = Storage.data() + pos;
+        }
+
+        std::for_each(Indexes.cbegin(), Indexes.cend(), [&](ui32 index) { Fields[index] = static_cast<NUdf::TUnboxedValue*>(ptr++); });
+    }
+
+public:
+    TSpillingSupportState(TMemoryUsageInfo* memInfo, ui64 count, const bool* directons, size_t keyWidth, const TCompareFunc& compare,
+        const std::vector<ui32>& indexes, IComputationWideFlowNode *const flow, TMultiType* tupleMultiType)
+        : TBase(memInfo)
+        , Flow(flow)
+        , Count(count)
+        , Indexes(indexes)
+        , Directions(directons, directons + keyWidth)
+        , LessFunc(std::bind(std::less<int>(), std::bind(compare, Directions.data(), std::placeholders::_1, std::placeholders::_2), 0))
+        , Fields(Indexes.size(), nullptr)
+        , TupleMultiType(tupleMultiType)
+    {
+        if constexpr (!HasCount) {
+            ResetFields();
+            return;
+        }
+        throw yexception() << "Spilling doesn't support TopSort.";
+    }
+
+    virtual EFetchResult DoCalculate(TComputationContext& ctx, NUdf::TUnboxedValue*const* output) {
+        while (true) {
+            switch(GetMode()) {
+                case EOperatingMode::InMemory: {
+                    auto r = DoCalculateInMemory(ctx, output);
+                    if (GetMode() == TSpillingSupportState::EOperatingMode::InMemory) {
+                        return r;
+                    }
+                    break;
+                }
+                case EOperatingMode::Spilling: {
+                    DoCalculateWithSpilling(ctx);
+                    if (GetMode() == EOperatingMode::Spilling) {
+                        return EFetchResult::Yield;
+                    }
+                    break;
+                }
+                case EOperatingMode::ProcessSpilled: {
+                    return ProcessSpilledData(output);
+                }
+
+            }
+        }
+        Y_UNREACHABLE();
+    }
+
+private:
+
+    EFetchResult DoCalculateInMemory(TComputationContext& ctx, NUdf::TUnboxedValue*const* output) {
+        while (EFetchResult::Finish != InputStatus) {
+            switch (InputStatus = Flow->FetchValues(ctx, GetFields())) {
+                case EFetchResult::One:
+                    if (Put()) {
+                        if (!HasMemoryForProcessing()) {
+                            SwitchMode(EOperatingMode::Spilling, ctx);
+                            return EFetchResult::Yield;
```

Comment:
Is it safe to return Yield here? It may cause a deadlock in case if no spilling operation started so far.

Reply:
Yes, it's safe. The code will go then to `DoCalculateWithSpilling` from `DoCalculate`